src/server/lock_manager/waiter_manager.rs

```diff
@@ -568,84 +592,92 @@ impl WaiterManager {
     fn handle_deadlock(
         &mut self,
         waiter_ts: TimeStamp,
-        lock: Lock,
+        key: Vec<u8>,
+        lock: LockDigest,
         deadlock_key_hash: u64,
         wait_chain: Vec<WaitForEntry>,
     ) {
-        if let Some(mut waiter) = self.wait_table.borrow_mut().remove_waiter(lock, waiter_ts) {
-            waiter.deadlock_with(deadlock_key_hash, wait_chain);
-            waiter.notify();
+        let waiter = self
+            .wait_table
+            .borrow_mut()
+            .take_waiter_by_lock_digest(lock, waiter_ts);
+        if let Some(waiter) = waiter {
+            waiter.cancel_for_deadlock(lock, key, deadlock_key_hash, wait_chain);
         }
     }
 
-    fn handle_config_change(
-        &mut self,
-        timeout: Option<ReadableDuration>,
-        delay: Option<ReadableDuration>,
-    ) {
+    fn handle_config_change(&mut self, timeout: Option<ReadableDuration>) {
         if let Some(timeout) = timeout {
             self.default_wait_for_lock_timeout = timeout;
         }
-        if let Some(delay) = delay {
-            self.wake_up_delay_duration = delay;
-        }
         info!(
             "Waiter manager config changed";
             "default_wait_for_lock_timeout" => self.default_wait_for_lock_timeout.to_string(),
-            "wake_up_delay_duration" => self.wake_up_delay_duration.to_string()
         );
     }
 }
 
 impl FutureRunnable<Task> for WaiterManager {
     fn run(&mut self, task: Task) {
         match task {
+            Task::SetKeyWakeUpDelayCallback { cb } => {
+                self.wait_table
+                    .borrow_mut()
+                    .set_wake_up_key_delay_callback(Some(cb));
+            }
             Task::WaitFor {
+                token,
+                region_id,
+                region_epoch,
+                term,
                 start_ts,
-                cb,
-                pr,
-                lock,
+                wait_info,
                 timeout,
+                cancel_callback,
                 diag_ctx,
                 start_waiting_time,
             } => {
                 let waiter = Waiter::new(
+                    region_id,
+                    region_epoch,
+                    term,
                     start_ts,
-                    cb,
-                    pr,
-                    lock,
+                    wait_info,
+                    cancel_callback,
                     self.normalize_deadline(timeout),
                     diag_ctx,
                     start_waiting_time,
                 );
-                self.handle_wait_for(waiter);
+                self.handle_wait_for(token, waiter);
                 TASK_COUNTER_METRICS.wait_for.inc();
             }
-            Task::WakeUp {
-                lock_ts,
-                hashes,
-                commit_ts,
-            } => {
-                self.handle_wake_up(lock_ts, hashes, commit_ts);
+
+            Task::RemoveLockWait { token } => {
```

Comment:
Do we have tests to cover the removing triggered by different reasons or it's planned in the next PRs?

Reply:
Currently a waiter in `WaiterManager` may only be removed in following cases:

* Receiving an `RemoveLockWait` message outside (because the lock-wait-entry is popped out from the queue and woken up, thus `LockWaitContext.finish_request` is called)
* Canceled due to timeout.
* Canceled due to deadlock

Canceling by region changes is not supported in this PR.
But I just noticed that `RemoveLockWait` is not covered in tests. I'll add it later.